#164139782 Pagination Support for Recipes

LordUche opened the pull request:
#### What does this PR do?
- adds pagination support for listing recipes
#### Description of Task to be completed?
- write unit tests
- accept `page` and `limit` query parameters from request url
- validate query parameters with `Joi`
- document feature with swagger
#### How should this be manually tested?
- ensure build is passing on Travis
- send a *GET* request to `/api/recipes` and pass in `page` and/or `limit` query params
#### Any background context you want to provide?
Pagination support is lacking for our recipes. Reading can be a nightmare when we have so many recipes.
#### What are the relevant pivotal tracker stories?
[#164139782](https://www.pivotaltracker.com/story/show/164139782)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

LordUche committed:
feat(pagination): add pagination functionality
[Finishes #164139782]

LordUche committed:
ft(recipes-pagination): document with swagger

LordUche committed:
feat(recipes-pagination): fix logic

andela-pessien reviewed (commented):
I think I have expressed my issue with including and pushing changes from unmerged feature branches in one's own feature branch more than often enough.

As it stands I cannot approve this PR because it contains changes I have commented on in a different PR and unless corrections based on the feedback given reflect here, it cannot get merged. I can't even review it properly because far too much of it is not your own work.

andela-pessien commented:
@LordUche Please update this - the corresponding PR has been merged

LordUche commented:
@andela-pessien I needed to have the `list recipes` functionality working for pagination to work. That was why I branched out of that feature branch. I have rebased with `develop` successfully and dropped all commits from the feature branch now.

andela-pessien commented on server/controllers/recipe-controller.js:
```diff
@@ -148,20 +152,30 @@ class RecipeController {
-  static getRecipes(req, res) {
-    Recipe.findAll(defaultRecipeDbFilter)
+  static async getRecipes(req, res) {
+    const { page } = req.query;
```
This doesn't match the spec's requirements - this endpoint should accept an _offset_, not a _page_.

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -141,4 +158,156 @@ describe('Recipes', () => {
+    before(async () => {
+      const array = Array(50)
+        .fill()
+        .map(() => {
```
Please address this

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -333,4 +331,115 @@ describe('Recipes', () => {
+          expect(res).to.have.status(200);
+          expect(recipes).to.be.an('Array');
+          expect(recipes.length).to.equal(10);
+          expect(recipes[0].id).to.not.equal(items[0].id);
```
You should test positives, not just negatives. (i.e. the fact that the first returned recipe is not the first created recipe does not guarantee whether the right recipes were returned or not). That applies to the rest of the examples after this here as well

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -333,4 +331,115 @@ describe('Recipes', () => {
+        .get(url)
+        .end((err, res) => {
+          const { recipes } = res.body;
+          items = recipes;
```
You shouldn't be performing setup for other examples inside a test example, as the examples in a suite should be runnable in _any_ order (or even in isolation). That's what the pre- and post-hooks are for.

LordUche committed:
feat(recipes-pagination): use offset instead of page

LordUche committed:
feat(recipes-pagination): fix test

LordUche committed:
feat(recipes-pagination): test for offsets
